### PR TITLE
Move job's low level methods to tests only

### PR DESCRIPTION
**Pull request by torymur**
- move low level methods to basic testcase, like:

```
project.start_job/client.start_job -> start_job,
job.finished -> expand to jobq.finish(),
job.failed -> expand to jobq.finish(),
```
- jobq._set_state -> jobq.update()
- add tests for jobq.update()

### CODE REVIEW

File: tests/test_client.py [hunk 1]
```diff
@@ -22,36 +22,6 @@ def test_push_job(self):
         self.assertEqual(m.get('state'), u'deleted')
         self.assertEqual(m.get('foo'), u'baz')
 
-    def test_botgroup(self):
```
Comment (dangra) [inline tests/test_client.py]: why did you remove this test case?

File: tests/test_project.py [hunk 2]
```diff
@@ -169,7 +169,7 @@ def test_requests(self):
         r3 = job.requests.add(url='http://test.com/3', status=400, method='PUT',
                               rs=0, duration=1, parent=r1, ts=ts + 2, fp='1234')
 
-        job.close_writers()
+        self.close_job_writers(job)
```
Comment (dangra) [inline tests/test_project.py]: can be replaced by `job.requests.close()`

File: tests/test_project.py [hunk 3]
```diff
@@ -121,12 +121,12 @@ def test_broad(self):
...
         job = project.push_job(self.spidername)
         self.assertEqual(job.metadata.get('state'), 'pending')
-        job = project.start_job()
+        job = self.start_job()
         self.assertEqual(job.metadata.get('state'), 'running')
         job.items.write({'title': 'bar'})
         job.logs.info('nice to meet you')
         job.samples.write([1, 2, 3])
-        job.finished()
+        self.job_finished(job)
```
Comment (dangra) [inline tests/test_project.py]: This is the only place where `close_writers()` makes sense and calling `self.hsc.close()` to close writers is not an option.

the alternative is to close each writer manually:

    job.items.close()
    job.logs.close()
    job.samples.close()
    self.hsclient.jobq.finish(job)

File: tests/test_system.py [hunk 4]
```diff
@@ -64,12 +64,12 @@ def _run_runner(self, pushed, close_reason):
             try:
                 self._run_scraper(job.key, job.jobauth, close_reason=close_reason)
             except Exception as exc:
-                job.failed(message=str(exc))
+                self.job_failed(job, message=str(exc))
```
Comment (dangra) [inline tests/test_system.py]: you can expand job_failed by its implementation here.

File: tests/test_system.py [hunk 5]
```diff
@@ -64,12 +64,12 @@ def _run_runner(self, pushed, close_reason):
...
                 # logging from runner must append and never remove messages logged
                 # by scraper
                 self.assertTrue(job.logs.batch_append)
             else:
-                job.finished()
+                self.job_finished(job)
```
Comment (dangra) [inline tests/test_system.py]: And this can be replaced by `client.jobq.finish(job)`

File: hubstorage/job.py [hunk 6]
```diff
@@ -20,15 +20,6 @@ def __init__(self, client, key, auth=None, jobauth=None, metadata=None):
         self.requests = Requests(client, self.key, self.auth)
         self.jobq = JobQ(client, self.key.split('/')[0], auth)
 
-    def close_writers(self):
```
Comment (dangra) [inline hubstorage/job.py]: I think this method has a place in Job class. It allows scraper writers to correctly close the writers without a need to close all batchuploader writers.

File: tests/hstestcase.py [hunk 7]
```diff
@@ -71,6 +71,33 @@ def _unset_testbotgroup(cls):
...
         cls.project.settings.expire()
 
+    def start_job(self, **startparams):
+        jobdata = self.project.jobq.start(**startparams)
+        if jobdata:
+            jobkey = jobdata.pop('key')
+            jobauth = (jobkey, jobdata['auth'])
+            return self.project.get_job(jobkey, jobauth=jobauth, metadata=jobdata)
+
+    def job_finished(self, job, close_reason=None):
```
Comment (dangra) [inline tests/hstestcase.py]: can be replaced in most places directly by call to `self.hsclient.jobq.finish(job)`

File: tests/hstestcase.py [hunk 8]
```diff
@@ -71,6 +71,33 @@ def _unset_testbotgroup(cls):
...
+        close_reason = close_reason or \
+            job.metadata.liveget('close_reason') or 'no_reason'
+        self.close_job_writers(job)
+        job.jobq.finish(job, close_reason=close_reason)
+
+    def job_failed(self, job, reason='failed', message=None):
```
Comment (dangra) [inline tests/hstestcase.py]: same than job_finished and only used by test_system

File: tests/test_client.py (hunk 1 above)
Comment (torymur) [inline tests/test_client.py]: Good question, it was mostly about the fact, that we are testing test method. And this test to pass requires different implementation of ``start_job`` (because default ``project.jobq`` starts job no matter was botgroup defined or not). So, current simplified version of ``start_job``, which uses only project's jobq, wouldn't be enough and we need to go back to our old good version of ``client.start_job`` with checking did we receive projectid or not (only maybe without auth argument). From that perspective seems it's 'too much' for testing test method. But I'm ok with keeping it too, of course :) So, what do you think?

Comment (torymur) [inline tests/test_client.py]: In keeping case we can save the other one as well - ``test_start_job``. Only fix assertion ``        self.assertEqual(dict(j1.metadata), dict(j2.metadata))``, because it would fail as ``start_job`` returns auth key and ``get_job`` not. So with something like ``j1.metadata.pop('auth')``.

File: tests/test_system.py (hunk 5 above)
Comment (torymur) [inline tests/test_system.py]: ok, will do it, thanks for all your notes here!

File: tests/test_client.py (hunk 1 above)
Comment (dangra) [inline tests/test_client.py]: I understand, the same is tested in test_jobq.py so it is fine to remove it.
